pkg/volume/csi/csi_test.go

```diff
@@ -0,0 +1,366 @@
+				if err != nil {
+					t.Log("csiTest.VolumeAll failed to create new device mounter: ", err)
+				}
+			}
+
+			if devMounter != nil {
+				csiDevMounter := devMounter.(*csiAttacher)
+				csiDevMounter.csiClient = csiClient
+				devMountPath, err := csiDevMounter.GetDeviceMountPath(volSpec)
+				if err != nil {
+					t.Fatalf("csiTest.VolumeAll deviceMounter.GetdeviceMountPath failed %s", err)
+				}
+				if err := csiDevMounter.MountDevice(volSpec, devicePath, devMountPath); err != nil {
+					t.Fatalf("csiTest.VolumeAll deviceMounter.MountDevice failed: %v", err)
+				}
+				t.Log("csiTest.VolumeAll device mounted at path:", devMountPath)
+			} else {
+				t.Log("csiTest.VolumeAll DeviceMountablePlugin not found, skipping deviceMounter.MountDevice")
+			}
+
+			// mount volume
+			t.Log("csiTest.VolumeAll Mouting volume...")
+			volPlug, err := plugMgr.FindPluginBySpec(volSpec)
+			if err != nil || volPlug == nil {
+				t.Fatalf("csiTest.VolumeAll PluginMgr.FindPluginBySpec failed: %v", err)
+			}
+
+			if volPlug == nil {
+				t.Fatalf("csiTest.VolumeAll volumePlugin is nil")
+			}
+
+			if !volPlug.CanSupport(volSpec) {
+				t.Fatal("csiTest.VolumeAll volumePlugin.CanSupport returned false")
+			}
+
+			mounter, err := volPlug.NewMounter(volSpec, pod, volume.VolumeOptions{})
+			if err != nil || mounter == nil {
+				t.Fatalf("csiTest.VolumeAll volPlugin.NewMounter is nil or error: %s", err)
+			}
+
+			if err := mounter.CanMount(); err != nil {
+				t.Log("csiTest.VolumeAll mounter.CanMount failed, skipping mount")
+				return
+			}
+
+			var fsGroup *int64
+			if pod.Spec.SecurityContext != nil && pod.Spec.SecurityContext.FSGroup != nil {
+				fsGroup = pod.Spec.SecurityContext.FSGroup
+			}
+
+			csiMounter := mounter.(*csiMountMgr)
+			csiMounter.csiClient = csiClient
+			if err := csiMounter.SetUp(fsGroup); err != nil {
+				t.Fatalf("csiTest.VolumeAll mounter.Setup(fsGroup) failed: %s", err)
+			}
+			t.Log("csiTest.VolumeAll mounter.Setup(fsGroup) done OK")
+
+			dataFile := filepath.Join(path.Dir(mounter.GetPath()), volDataFileName)
+			if _, err := os.Stat(dataFile); err != nil {
+				t.Fatalf("csiTest.VolumeAll meatadata JSON file not found: %s", dataFile)
+			}
+			t.Log("csiTest.VolumeAll JSON datafile generated OK:", dataFile)
+
+			// ******** Reconstruct volume if necessary ********* //
+			if test.shouldConstructVol {
+				volPath := path.Dir(csiMounter.GetPath())
+				t.Log("csiTest.VolumeAll entering plugin.ConstructVolumeSpec for path", volPath)
+				spec, err := volPlug.ConstructVolumeSpec(test.volName, volPath)
+				if err != nil {
+					t.Fatalf("csiTest.VolumeAll plugin.ConstructVolumeSpec failed: %s", err)
+				} else {
+					if volSpec == nil {
+						t.Logf("csiTest.VolumeAll plugin.ConstructVolumeSpec returned nil spec")
+					} else {
+						volSpec = spec
+					}
+				}
+			} else {
+				t.Log("csiTest.VolumeAll test.ShouldConstructVol == false, skipping plugin.ConstructVolumeSpec")
```

Comment:
What determines if a volume should or should not reconstruct?

Reply:
Right now, just the test flag.  I guess for the CSI case, reconstruction is always supported unless an error is returned.

Reply:
I removed the flag.